base.go

```diff
@@ -348,47 +348,56 @@ func getBuilt(o BaseLayerOpts, sfm StackerFiles) error {
 	// may not have been copied before and the final `umoci repack` expects
 	// them to be there.
 	targetName := o.Name
-	baseLayerDefinition, ok := sfm.LookupLayerDefinition(o.Layer.From.Tag)
-	if !ok {
-		return errors.Errorf("missing base layer: %s?", o.Layer.From.Tag)
-	}
+	base := o.Layer
+	var baseTag string
+	var baseType string
+
+	for {
+		// Iterate through base layers until we find the first one which is not BuiltType or BuildOnly
+
+		// Need to declare ok and err  separately, if we do it in the same line as
+		// assigning the new value to base, base would be a new variable only in the scope
+		// of this iteration and we never meet the condition to exit the loop
+		var ok bool
+		var err error
 
-	if baseLayerDefinition.BuildOnly {
-		baseInputTag, err := baseLayerDefinition.From.ParseTag()
+		baseTag, err = base.From.ParseTag()
 		if err != nil {
 			return err
 		}
+		baseType = base.From.Type
 
-		// Ok, we didn't generate an OCI diff for whatever mutated this
-		// layer. So either 1. the layer was sourced
-		// from something and is in the OCI cache, so we can copy at
-		// least the base's base over to make the delta smaller or 2.
-		// it was sourced from e.g. a tar file, in which case there's
-		// nothing we can do besides initialize an empty oci tag and
-		// generate the whole thing.
-		cacheDir := path.Join(o.Config.StackerDir, "layer-bases", "oci")
-		err = lib.ImageCopy(lib.ImageCopyOpts{
-			Src:  fmt.Sprintf("oci:%s:%s", cacheDir, baseInputTag),
-			Dest: fmt.Sprintf("oci:%s:%s", o.Config.OCIDir, targetName),
-		})
-		if err != nil {
-			// give up and just create an empty image
-			return umociInit(o)
+		if baseType != BuiltType {
+			break
 		}
-		return nil
-	} else {
-		baseOutputTag, err := o.Layer.From.ParseTag()
-		if err != nil {
-			return err
+
+		base, ok = sfm.LookupLayerDefinition(base.From.Tag)
+		if !ok {
+			return fmt.Errorf("missing base layer: %s?", base.From.Tag)
+		}
+
+		if !base.BuildOnly {
+			break
 		}
+	}
 
+	if baseType != DockerType && baseType != OCIType && baseType != ZotType {
+		// Assume the user who wrote the stacker yaml has ordered the layers correctly
+		// and the base image has already been built and placed in  OCIDir
 		refs, _ := o.OCI.ListReferences(context.Background())
 		fmt.Printf("refs: %v\n", refs)
 		return lib.ImageCopy(lib.ImageCopyOpts{
```

Comment:
If the layer is build-only, this fails, right?

Reply:
If the only layers which can be 'build-only' are BuiltType, then this should be covered.
The look should break only if base is not BuiltType (line 370) or it is BuiltType but it is not 'build-only' (line 379)